### PR TITLE
Test installation on macOS using GitHub-hosted runners

### PR DESCRIPTION
**Pull request by lastoCHka42**
Added 'host mode' as CLI argument to run the check on the host OS
(neither VM nor docker). In this case, we don't need config file and can
set necessary parameters (tarantool version, build) as CLI arguments.
Use case: CI mode, check the macOS build on the runner (we can't run osx
on docker or VM).

Resolves https://github.com/tarantool/delivery-checker/issues/65

### CODE REVIEW

**Comment by lastoCHka42** [comment]
Testing check list:

- [ ] Tests for MacOS are OK
- [ ] Tests for other OS versions are OK
- [ ] VK Teams notifications are OK

File: build_tester/builders/host_builder.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+class HostBuilder:
+    def __init__(
+            self,
+            build_info: HostInfo,
+            archive_dir_path='./archive',
+            scripts_dir_path='.',
+            prepare_dir_path='./prepare',
+            tests_dir_path='./local/tests',
+            results_file_path='results.json',
+            log_func=print):
```
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply build_tester/builders/host_builder.py:20]: The same as it was in DockerInfo and VirtualboxInfo

File: build_tester/builders/host_builder.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            os.environ["TNT_VERSION"] = self.build_info.tarantool_version
+            os.environ["RESULTS_FILE"] = self.results_file_path
+            os.system(os.path.join(str(self.scripts_dir_path), 'init.lua'))
+
+            # Get the path to result file where the class 'Result' awaits it
+            path = os.path.join(str(self.tests_dir_path),
+                                f'{self.build_info.os_name}_'
+                                f'{self.build_info.build_name}.json')
+            os.system(f'mv {self.results_file_path} {path}')
+        except Exception:
```
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply build_tester/builders/host_builder.py:52]: I din't change this part, just put it under `if`

File: config/config.py [hunk 3]
```diff
@@ -105,48 +132,58 @@ def __init__(self, cli_args, config_json):
...
-        self.virtual_box_params = {
-            k: v['virtual_box']
-            for k, v in os_params.items()
-            if v.get('virtual_box') is not None and (not self.dist or self.dist == k)
-        }
-        self.json = config_json
+        if not self.host_mode:
+            os_params = config_json.get('os_params')
+            assert config_json.get(
+                'os_params') is not None, 'No OS params in config!'
```
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply config/config.py:163]: I din't change this part, just put it under if

File: build_tester/builders/host_builder.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply build_tester/builders/host_builder.py:52]: done

File: config/config.py [hunk 4]
```diff
@@ -23,6 +24,22 @@
 }
 
 
+def get_host_os_version():
```
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply config/config.py:27]: done

**Review by ochaplashkin — APPROVED** [review]
LGTM

File: build_tester/builders/host_builder.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import os
+import traceback
+from collections import namedtuple
```
Comment (ylobankov) [inline build_tester/builders/host_builder.py:3]: Looks like imports are not in alphabet order.

File: build_tester/builders/host_builder.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            results_file_path='results.json'):
+        self.build_info = build_info
+        self.archive_dir_path = archive_dir_path
+        self.scripts_dir_path = scripts_dir_path
+        self.prepare_dir_path = prepare_dir_path
+        self.tests_dir_path = tests_dir_path
+        self.results_file_path = results_file_path
+
+    def run_check(self):
+
```
Comment (ylobankov) [inline build_tester/builders/host_builder.py:29]: Let's remove this empty line.

File: build_tester/builders/host_builder.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            tests_dir_path='./local/tests',
+            results_file_path='results.json'):
+        self.build_info = build_info
+        self.archive_dir_path = archive_dir_path
+        self.scripts_dir_path = scripts_dir_path
+        self.prepare_dir_path = prepare_dir_path
+        self.tests_dir_path = tests_dir_path
+        self.results_file_path = results_file_path
+
+    def run_check(self):
```
Comment (ylobankov) [inline build_tester/builders/host_builder.py:28]: Shouldn't it be named just `run`?

File: build_tester/tester.py [hunk 8]
```diff
@@ -6,6 +6,7 @@
 import requests
 
 from build_tester.builders.docker_builder import DockerBuilder, DockerInfo
+from build_tester.builders.host_builder import HostInfo, HostBuilder
```
Comment (ylobankov) [inline build_tester/tester.py:9]: Shouldn't it be in this order: `HostBuilder, HostInfo`.

File: build_tester/tester.py [hunk 9]
```diff
@@ -83,6 +84,21 @@ def __download_scripts(self):
         with open(os.path.join(self.config.install_dir_path, 'default.sh'), mode='r') as fs:
             default_script = fs.read()
 
+        if self.config.host_mode:
+            os_name = 'macos' if self.config.dist == 'os-x' else self.config.dist
```
Comment (ylobankov) [inline build_tester/tester.py:88]: Why not just `osx`?

File: config/config.py [hunk 10]
```diff
@@ -1,6 +1,7 @@
 import os
 from dataclasses import dataclass
-from typing import Optional
+import platform
+from datetime import datetime
```
Comment (ylobankov) [inline config/config.py:4]: Not alphabet order.

File: config/config.py [hunk 11]
```diff
@@ -76,23 +95,29 @@ class CheckerConfig:
...
+                self.dist_version = \
+                    debian_version_to_name.get(cli_args.dist_version) or \
+                    cli_args.dist_version or None
+            else:
+                self.dist_version = cli_args.dist_version or None
 
         self.console_mode = cli_args.console_mode
         self.debug_mode = cli_args.debug_mode
-        self.ci_mode = cli_args.ci_mode or os.environ.get('GITHUB_ENV') is not None
+        self.ci_mode = cli_args.ci_mode or os.environ.get(
```
Comment (ylobankov) [inline config/config.py:119]: Unnecessary change.

File: config/config.py [hunk 12]
```diff
@@ -128,28 +153,31 @@ def __init__(self, cli_args, config_json):
...
+            self.send_to_remote = config_json.get('send_to_remote', {})
+            self.use_remote_results = config_json.get('use_remote_results', False)
+
+            self.docker_params = {
+                k: v['docker']
+                for k, v in os_params.items()
+                if v.get('docker') is not None and (not self.dist or self.dist == k)
+            }
+            if self.dist_version:
+                assert self.dist_version in self.docker_params[self.dist][
```
Comment (ylobankov) [inline config/config.py:169]: Please return formatting as it was before.

File: scripts/prepare/os-x.sh [hunk 13]
```diff
@@ -3,11 +3,9 @@
...
 
-# Preparation commands
-
 # Download dependencies to cache it
 brew install --only-dependencies tarantool
 brew install --only-dependencies tarantool --HEAD
 
 # Shutdown VM to indicate that preparation is finished
-sudo shutdown -h now || true
+#sudo shutdown -h now || true
```
Comment (ylobankov) [inline scripts/prepare/os-x.sh:11]: ?

File: build_tester/builders/host_builder.py (hunk 5 above)
Comment (lastoCHka42) [reply build_tester/builders/host_builder.py:3]: I tried to keep consistency with other files

File: build_tester/builders/host_builder.py (hunk 7 above)
Comment (lastoCHka42) [reply build_tester/builders/host_builder.py:28]: fixed

File: build_tester/builders/host_builder.py (hunk 6 above)
Comment (lastoCHka42) [reply build_tester/builders/host_builder.py:29]: fixed

File: build_tester/tester.py (hunk 8 above)
Comment (lastoCHka42) [reply build_tester/tester.py:9]: fixed

File: build_tester/tester.py (hunk 9 above)
Comment (lastoCHka42) [reply build_tester/tester.py:88]: Because on the Website API (where we get the installation commands) we use `os-x`

File: config/config.py (hunk 10 above)
Comment (lastoCHka42) [reply config/config.py:4]: fixed

File: config/config.py (hunk 11 above)
Comment (lastoCHka42) [reply config/config.py:119]: fixed

File: scripts/prepare/os-x.sh (hunk 13 above)
Comment (lastoCHka42) [reply scripts/prepare/os-x.sh:11]: fixed

File: config/config.py (hunk 12 above)
Comment (lastoCHka42) [reply config/config.py:169]: fixed

File: config/config.py [hunk 14]
```diff
@@ -34,41 +34,44 @@ class CheckerConfig:
...
+    scripts_dir_path: str  # Dir name for scripts. Config, default './scripts'
+    prepare_dir_name: str  # Dir name with bash scripts to prepare OS, config, default 'prepare'
+    prepare_dir_path: str  # Path to prepare dir, config or './scripts/prepare'
+    install_dir_name: str  # Dir name for installation instructions, config or 'install'
+    install_dir_path: str  # Path to installation dir, config or './scripts/install'
+    local_dir_path: str  # Path to archive dir in VM or container. Config or './local'
+    remote_dir_path: str  # Config or './remote'
+    archive_dir_path: str  # Path to save check logs and test result, config or './archive'
+    logs_dir_name: str  # Name for check logs dir, config or 'logs'
+    logs_dir_path: str  # Path to check logs dir on VM or container, config ot './local/logs'
```
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply config/config.py]: fixed

File: .github/workflows/osx.yml [hunk 15]
```diff
@@ -0,0 +1,57 @@
+name: MacOS
```
Comment (ylobankov) [inline .github/workflows/osx.yml:1]: macOS

File: .github/workflows/osx.yml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+on:
+  schedule:
+    - cron: '0 2 * * *'
+  workflow_dispatch:
+  pull_request:
+    paths:
+      - '.github/workflows/os-x.yml'
```
Comment (ylobankov) [inline .github/workflows/osx.yml:8]: os-x -> osx

File: .github/workflows/osx.yml [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    (
+    github.ref == 'refs/heads/master' ||
+    startsWith(github.ref, 'refs/tags/')
+    ) &&
+    format('{0}-{1}', github.run_id, github.run_attempt) ||
+    format('{0}-{1}', github.workflow, github.ref) }}
+  cancel-in-progress: true
+
+jobs:
+  os-x:
```
Comment (ylobankov) [inline .github/workflows/osx.yml:32]: osx

File: .github/workflows/amazon-aarch64.yml [hunk 18]
```diff
@@ -59,3 +59,10 @@ jobs:
         with:
           name: ${{ env.NOW }}-amazon-aarch64-${{ github.run_id }}
           path: archive
+
+      - name: Send notification
+        if: failure()
+        uses: ./.github/actions/failure_notify
+        with:
+          bot-token: ${{ secrets.VKTEAMS_TARANTOOLBOT_TOKEN }}
+          chat-id: ${{ secrets.VKTEAMS_SERVICE_CHAT_ID }}
```
Comment (ylobankov) [inline .github/workflows/amazon-aarch64.yml:68]: Will this work?

File: .github/workflows/amazon-aarch64.yml [hunk 19]
```diff
@@ -59,3 +59,10 @@ jobs:
         with:
           name: ${{ env.NOW }}-amazon-aarch64-${{ github.run_id }}
           path: archive
+
+      - name: Send notification
+        if: failure()
+        uses: ./.github/actions/failure_notify
```
Comment (ylobankov) [inline .github/workflows/amazon-aarch64.yml:65]: Usually, we use dash instead of underscore.

Comment (lastoCHka42) [reply .github/workflows/amazon-aarch64.yml:65]: fixed

File: .github/workflows/amazon-aarch64.yml (hunk 18 above)
Comment (lastoCHka42) [reply .github/workflows/amazon-aarch64.yml:68]: fixed)

File: .github/actions/failure-notify/action.yml [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    required: false
+    default: 'https://api.internal.myteam.mail.ru/bot/v1/'
+  service-chat-id:
+    description: 'VK Teams chat for service notification (failure on prod)'
+    required: true
+  debug-chat-id:
+    description: 'VK Teams chat for debug notification (failure on unstable branches)'
+    required: true
+runs:
+  using: "composite"
```
Comment (ylobankov) [inline .github/actions/failure-notify/action.yml:19]: You can omit quotes.

File: .github/workflows/amazon-aarch64.yml [hunk 21]
```diff
@@ -56,6 +56,7 @@ jobs:
         run: echo "NOW=$(date +'%Y-%m-%d-%H-%M')" >> $GITHUB_ENV
 
       - uses: actions/upload-artifact@v3
+        if: always()
```
[the remark this answers is not in the document]
Comment (lastoCHka42) [reply .github/workflows/amazon-aarch64.yml:60]: fixed

File: .github/actions/failure-notify/action.yml (hunk 20 above)
Comment (lastoCHka42) [reply .github/actions/failure-notify/action.yml]: done

**Review by ylobankov — APPROVED** [review]
LGTM
